src/coreclr/System.Private.CoreLib/src/System/Array.CoreCLR.cs

```diff
@@ -554,8 +413,98 @@ private unsafe nint GetFlattenedIndex(ReadOnlySpan<int> indices)
+                    else
+                        SpanHelpers.ClearWithoutReferences(ref offsetDataRef, pElementMethodTable->GetNumInstanceFieldBytes());
+                }
+                else
+                {
+                    ref object? elementRef = ref Unsafe.As<byte, object?>(ref arrayDataRef);
+                    ref object? offsetElementRef = ref Unsafe.Add(ref elementRef, (nuint)flattenedIndex);
+                    offsetElementRef = null;
+                }
+            }
+            else if (TypeHandle.AreSameType(arrayElementTypeHandle, TypeHandle.TypeHandleOf<object>()))
+            {
+                // Everything is compatible with Object
+                ref object? elementRef = ref Unsafe.As<byte, object?>(ref arrayDataRef);
+                ref object? offsetElementRef = ref Unsafe.Add(ref elementRef, (nuint)flattenedIndex);
+                offsetElementRef = value;
+            }
+            else if (!pElementMethodTable->IsValueType)
+            {
+                if (CastHelpers.IsInstanceOfAny(pElementMethodTable, value) == null)
+                    throw new InvalidCastException(SR.InvalidCast_StoreArrayElement);
+
+                ref object? elementRef = ref Unsafe.As<byte, object?>(ref arrayDataRef);
+                ref object? offsetElementRef = ref Unsafe.Add(ref elementRef, (nuint)flattenedIndex);
+                offsetElementRef = value;
+            }
+            else
+            {
+                // value class or primitive type
+
+                ref byte offsetDataRef = ref Unsafe.Add(ref arrayDataRef, flattenedIndex * pMethodTable->ComponentSize);
+                if (CastHelpers.IsInstanceOfAny(pElementMethodTable, value) != null)
+                {
+                    if (pElementMethodTable->IsNullable)
+                    {
+                        RuntimeHelpers.Unbox_Nullable(ref offsetDataRef, pElementMethodTable, value);
+                    }
+                    else if (pElementMethodTable->ContainsGCPointers)
+                    {
+                        Buffer.BulkMoveWithWriteBarrier(ref offsetDataRef, ref value.GetRawData(), pElementMethodTable->GetNumInstanceFieldBytes());
+                    }
+                    else
+                    {
+                        SpanHelpers.Memmove(ref offsetDataRef, ref value.GetRawData(), pElementMethodTable->GetNumInstanceFieldBytes());
+                    }
+                }
+                else
+                {
+                    // Allow enum -> primitive conversion, disallow primitive -> enum conversion
+                    MethodTable* thSrc = RuntimeHelpers.GetMethodTable(value);
+                    CorElementType srcType = thSrc->GetVerifierCorElementType();
+                    CorElementType targetType = pElementMethodTable->GetVerifierCorElementType();
+
+                    if (!srcType.IsPrimitiveType() || !targetType.IsPrimitiveType() || !pElementMethodTable->IsTruePrimitive)
```

Comment:
```suggestion
                    // Array.SetValue() does *not* permit conversion from a primitive to an Enum.  
                    if (!pValueMethodTable->IsPrimitive || !pElementMethodTable->IsTruePrimitive)
```
Would this work?

Reply:
Well `CorElementType` is still needed for deciding primitive widen in this case. This is also the case where `MethodTable.GetVerifierCorElementType` can be convenient.